SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift

```diff
@@ -0,0 +1,250 @@
+                owsFail("\(self.logTag) in \(#function) unexpected dependency: \(dependency)")
+                continue
+            }
+
+            self.registeredRecipientIds.formUnion(batchOperation.registeredRecipientIds)
+        }
+
+        self.reportSuccess()
+    }
+}
+
+class LegacyContactDiscoveryBatchOperation: OWSOperation {
+
+    var registeredRecipientIds: Set<String>
+
+    private let recipientIdsToLookup: [String]
+    private var networkManager: TSNetworkManager {
+        return TSNetworkManager.shared()
+    }
+
+    // MARK: Initializers
+
+    required init(recipientIdsToLookup: [String]) {
+        self.recipientIdsToLookup = recipientIdsToLookup
+        self.registeredRecipientIds = Set()
+
+        super.init()
+
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+    }
+
+    // MARK: OWSOperation Overrides
+
+    // Called every retry, this is where the bulk of the operation's work should go.
+    override func run() {
+        Logger.debug("\(logTag) in \(#function)")
+
+        var phoneNumbersByHashes: [String: String] = [:]
+
+        for recipientId in recipientIdsToLookup {
+            let hash = Cryptography.truncatedSHA1Base64EncodedWithoutPadding(recipientId)
+            phoneNumbersByHashes[hash] = recipientId
+        }
+
+        let hashes: [String] = Array(phoneNumbersByHashes.keys)
+
+        let request = OWSRequestFactory.contactsIntersectionRequest(withHashesArray: hashes)
+
+        self.networkManager.makeRequest(request,
+                                        success: { (task, responseDict) in
+                                            do {
+                                                self.registeredRecipientIds = try self.parse(response: responseDict, phoneNumbersByHashes: phoneNumbersByHashes)
+                                                self.reportSuccess()
+                                            } catch {
+                                                self.reportError(error)
+                                            }
+        },
+                                        failure: { (task, error) in
+                                            guard let response = task.response as? HTTPURLResponse else {
+                                                let responseError: NSError = OWSErrorMakeUnableToProcessServerResponseError() as NSError
+                                                responseError.isRetryable = true
+                                                self.reportError(responseError)
+                                                return
+                                            }
+
+                                            guard response.statusCode != 413 else {
+                                                let rateLimitError = OWSErrorWithCodeDescription(OWSErrorCode.contactsUpdaterRateLimit, "Contacts Intersection Rate Limit")
```

Comment:
FYI this response will eventually include a hint about when to retry.

Reply:
Wait, I meant the CDS rate limit errors.

Reply:
👍